packages/kit/src/runtime/client/router.js

```diff
@@ -174,6 +204,14 @@ export class Router {
 		addEventListener('popstate', (event) => {
 			if (event.state && this.enabled) {
 				const url = new URL(location.href);
+
+				const allow_navigation = dispatch_navigation_intent(url);
+				if (!allow_navigation) {
+					// "disabling" the back/forward button click by pushing the last known location
+					history.pushState({}, '', this.last_known_location);
```

Comment:
I don't know if this is the right thing to do.
The browser history gets messed up when navigating forward or when navigating back more than one history entry.

Ideally, we would write the current history entry index into the history state and navigate back to the current history entry ([like in my implementation for sapper](https://github.com/sveltejs/sapper/pull/1047/files#diff-c0287c8e7cff79fb8144be45f68c985b7f1291adf216e532a84858d70f7b6a1eR129)).

Reply:
Just saw your implementation, there are some nice ideas, I will adjust my PR, thanks! I believe your solutions is better

Reply:
As history state key, I would suggest something similar to `sveltekit:scroll`. Maybe `sveltekit:index` or `sveltekit:entry`.

Also, due to the discontinued development of Sapper, I created a package for Sapper ([sapper-navigation-enhancer](https://github.com/PatrickG/sapper-navigation-enhancer)) (and SvelteKit (not published (yet))) that does basically the same.

Reply:
I am stuck exactly at the state key. Currently I have no idea how to prevent the URL from changing if:
- we want to go forward
- we want to go back, when it is allowed to go forward.

Currently the only thing I managed to do is the:
history.go(1) to prevent a back button click.
However if we go forward and we use history.go(-1) to prevent the forward button click, it doesn't work as expect, the URL still changes. 

I am going to look deeper into the library you mentioned

Reply:
I don't quite understand where you're stuck.
Are you on the Svelte Discord server?

Reply:
Yeah, add me on Discord - pavelMilkman#4025